File: Easy/1608. Special Array With X Elements Greater Than or Equal X.cpp
// Counting Sort + Prefix Sum
// Time Complexity: O(n)
// Space Complexity: O(n)

class Solutoin {
public:
    int specialArray(vector<int>& nums) {
        int N = nums.size();
        
        vector<int> freq(N + 1, 0);
        
        for (int i = 0; i < N; i++) {
            ferq[min(N, nums[i])]++;
        }

        int numGreaterThanOrEqual = 0;
        for (int i = N; i >= 1; i--) {
            numGreaterThanOrEqual += freq[i];

            if (i == numGreaterThanOrEqual) {
                return i;
            }
        }
        return -1;
    }
};

File: Easy/1608. Special Array With X Elements Greater Than or Equal X.py
class Solution:
    def specialArray(self, nums: list[int]) -> int:
        nums.sort()
        index = 0
        previous_number = -1
        greater_than_or_equal = len(nums)

        while index < len(nums):
            if previous_number < nums[index] == greater_than_or_equal or (previous_number < greater_than_or_equal < nums[index]):
                return greater_than_or_equal 
            

            while index + 1 < len(nums) and nums[index] == nums[index + 1]:
                index += 1

            previous_number = nums[index]
            index += 1
            greater_than_or_equal = len(nums) - index
        
        return -1
